src/lib/mixer/geoms/tools/px_generate_mixers.py

```diff
@@ -0,0 +1,356 @@
+    return torque
+
+def geom_to_torque_matrix(geom):
+    '''
+    Compute torque matrix Am and Bm from geometry dictionnary
+    Am is a 3xN matrix where N is the number of rotors
+    Each column is the torque generated by one rotor
+    '''
+    Am = torque_matrix(center=np.array([rotor['position'] for rotor in geom['rotors']]),
+                       axis=np.array([rotor['axis'] for rotor in geom['rotors']]),
+                       dirs=np.array([[1.0 if rotor['direction'] == 'CCW' else -1.0]
+                                      for rotor in geom['rotors']]),
+                       Ct=np.array([[rotor['Ct']] for rotor in geom['rotors']]),
+                       Cm=np.array([[rotor['Cm']] for rotor in geom['rotors']])).T
+    return Am
+
+def thrust_matrix(axis, Ct):
+    '''
+    Compute thrust generated by rotors
+    '''
+    # Normalize rotor axis
+    ax = axis / np.linalg.norm(axis, axis=1)[:, np.newaxis]
+    thrust = Ct * ax
+    return thrust
+
+def geom_to_thrust_matrix(geom):
+    '''
+    Compute thrust matrix At from geometry dictionnary
+    At is a 3xN matrix where N is the number of rotors
+    Each column is the thrust generated by one rotor
+    '''
+    At = thrust_matrix(axis=np.array([rotor['axis'] for rotor in geom['rotors']]),
+                       Ct=np.array([[rotor['Ct']] for rotor in geom['rotors']])).T
+
+    return At
+
+def geom_to_mix(geom):
+    '''
+    Compute combined torque & thrust matrix A and mix matrix B from geometry dictionnary
+
+    A is a 6xN matrix where N is the number of rotors
+    Each column is the torque and thrust generated by one rotor
+
+    B is a Nx6 matrix where N is the number of rotors
+    Each column is the command to apply to the servos to get
+    roll torque, pitch torque, yaw torque, x thrust, y thrust, z thrust
+    '''
+    # Combined torque & thrust matrix
+    At = geom_to_thrust_matrix(geom)
+    Am = geom_to_torque_matrix(geom)
+    A = np.vstack([Am, At])
+
+    # Mix matrix computed as pseudoinverse of A
+    B = np.linalg.pinv(A)
+
+    return A, B
+
+def normalize_mix_px4(B):
+    '''
+    Normalize mix for PX4
+    This is for compatibility only and should ideally not be used
+    '''
+    B_norm = np.linalg.norm(B, axis=0)
+    B_max = np.abs(B).max(axis=0)
+
+    # Same scale on roll and pitch
+    B_norm[0] = max(B_norm[0], B_norm[1]) / np.sqrt(B.shape[0] / 2.0)
+    B_norm[1] = B_norm[0]
+
+    # Scale yaw separately
+    B_norm[2] = B_max[2]
+
+    # Same scale on x, y
+    B_norm[3] = max(B_max[3], B_max[4])
+    B_norm[4] = B_norm[3]
+
+    # Scale z thrust separately
+    B_norm[5] = B_max[5]
+
+    # Normalize
+    B_norm[np.abs(B_norm) < 1e-3] = 1
+    B_px = (B / B_norm)
+
+    return B_px
+
+def generate_mixer_multirotor_header(geom_list, use_normalized_mix=False, use_6dof=False):
+    '''
+    Generate C header file with same format as multi_tables.py
+    TODO: rewrite using templates (see generation of uORB headers)
```

Comment:
We can use a jinja template here, like https://github.com/PX4/Firmware/blob/master/src/modules/systemlib/param/templates/px4_parameters.c.jinja

Reply:
Yes that would be cleaner. 
I was not sure whether the script should generate one big file (current behaviour), or one file per geometry (like uORB and params).
I decided to use the same output format as mixer_table.py (though writing to a buffer instead of stdout!), and I left templating for later.

Reply:
One big file is fine, I'd say. I'm ok with changing to a template in a later step.